iree/hal/vmla/op_kernels_generic.h

```diff
@@ -238,32 +238,62 @@ Status Finite::Execute(absl::Span<const T> src_buffer,
   return OkStatus();
 }
 
+template <typename T>
+void TransposeRecurse(absl::Span<const T> src_buffer, absl::Span<T> dst_buffer,
+                      ShapeSpan src_shape, ShapeSpan dst_shape,
+                      absl::Span<const int> src_strides,
+                      absl::Span<const int> dst_strides,
+                      absl::Span<const int32_t> perm, int rank, int dim_i,
+                      size_t src_base_offset, size_t dst_base_offset) {
+  // Two cases:
+  // -- dim_i < rank - 1: iterate on dim_i; set offsets and recurse on dim_i + 1
+  // -- dim_i = rank - 1: base case, fast copy with strides and offsets
+
+  if (dim_i < rank - 1) {
+    int recurse_dim_i = dim_i + 1;
+    for (size_t i = 0; i < dst_shape[dim_i]; ++i) {
+      size_t src_offset = src_base_offset + i * src_strides[perm[dim_i]];
+      size_t dst_offset = dst_base_offset + i * dst_strides[dim_i];
+      TransposeRecurse(src_buffer, dst_buffer, src_shape, dst_shape,
+                       src_strides, dst_strides, perm, rank, recurse_dim_i,
+                       src_offset, dst_offset);
+    }
+  } else {
+    for (size_t i = 0; i < dst_shape[dim_i]; ++i) {
+      size_t src_i = src_base_offset + i * src_strides[perm[dim_i]];
```

Comment:
the compiler may be doing this for you but you could try moving the stride lookup outside of the loop here (and above) as it's constant per dim_i